content/en/blog/_posts/2023-07-18-swap-beta1-graduation/index.md

```diff
@@ -0,0 +1,231 @@
+in effect.
+
+Prior to detailing the calculation of the swap limit, it is necessary to define the following terms:
+* `nodeTotalMemory`: The total amount of physical memory available on the node.
+* `totalPodsSwapAvailable`: The total amount of swap memory on the node that is available for use by Pods (some swap memory may be reserved for system use).
+* `containerMemoryRequest`: The container's memory request.
+
+Swap limitation is configured as:
+`(containerMemoryRequest / nodeTotalMemory) * totalPodsSwapAvailable`.
+
+In other words, the amount of swap that a container is able to use is proportionate to its
+memory request, the node's total physical memory and the total amount of swap memory on
+the node that is available for use by Pods.
+
+It is important to note that, for containers within Burstable QoS Pods, it is possible to
+opt-out of swap usage by specifying memory requests that are equal to memory limits.
+Containers configured in this manner will not have access to swap memory.
+
+## How does it work?
+There are a number of possible ways that one could envision swap use on a node.
+When swap is already provisioned and available on a node,
+[we have proposed](https://github.com/kubernetes/enhancements/blob/9d127347773ad19894ca488ee04f1cd3af5774fc/keps/sig-node/2400-node-swap/README.md#proposal)
+the kubelet should be able to be configured such that:
+- It can start with swap on.
+- It will direct the Container Runtime Interface to allocate zero swap memory
+  to Kubernetes workloads by default.
+- You can configure the kubelet to specify swap utilization for the entire
+  node.
+
+Swap configuration on a node is exposed to a cluster admin via the
+[`memorySwap` in the KubeletConfiguration](/docs/reference/config-api/kubelet-config.v1).
+As a cluster administrator, you can specify the node's behaviour in the
+presence of swap memory by setting `memorySwap.swapBehavior`.
+
+The kubelet employs the CRI (container runtime interface) API to direct the CRI to
+configure specific cgroup v2 parameters (such as `memory.swap.max`) in a manner that will
+enable the desired swap configuration for a container. The CRI is then responsible to
+write these settings to the container-level cgroup.
+
+## How can I monitor swap?
+A notable deficiency in the Alpha version was the inability to monitor and introspect swap
+usage. This issue has been addressed in the Beta version introduced in Kubernetes 1.28, which now
+provides the capability to monitor swap usage through several different methods.
+
+The beta version of kubelet now collects
+[node-level metric statistics](/docs/reference/instrumentation/node-metrics/),
+which can be accessed at the /metrics/resource and /stats/summary endpoints. This allows users to
+monitor swap usage and remaining swap memory when using LimitedSwap. Additionally, a
+machine_swap_bytes metric has been added to cadvisor to show the total physical swap capacity of the
```

Comment:
```suggestion
`machine_swap_bytes` metric has been added to cadvisor to show the total physical swap capacity of the
```
How does this work if you're using an experimental cAdvisor-less kubelet?

Reply:
All of the metrics / stats under `/metrics/resource` and `/stats/summary` provide information about the current usage (for both memory and swap), and cadvisor metrics expose the total machine memory/swap size.

In my PR to support swap I've also extended the CRI API so that in the future the CRI would provide more information about swap statistics.